src/libos/src/process/syscalls.rs

```diff
@@ -85,6 +86,154 @@ fn clone_file_actions_safely(fdop_ptr: *const FdOp) -> Result<Vec<FileAction>> {
     Ok(file_actions)
 }
 
+// Syscall handler for Glibc's spawn
+pub fn do_spawn2(
+    child_pid_ptr: *mut u32,
+    path: *const i8,
+    argv: *const *const i8,
+    envp: *const *const i8,
+    fa: *const SpawnFileActions,
+) -> Result<isize> {
+    check_mut_ptr(child_pid_ptr)?;
+    let path = clone_cstring_safely(path)?.to_string_lossy().into_owned();
+    let argv = clone_cstrings_safely(argv)?;
+    let envp = clone_cstrings_safely(envp)?;
+    let file_actions = clone_file_actions_from_fa_safely(fa)?;
+    let current = current!();
+    debug!(
+        "spawn: path: {:?}, argv: {:?}, envp: {:?}, actions: {:?}",
+        path, argv, envp, file_actions
+    );
+
+    let child_pid = super::do_spawn::do_spawn(&path, &argv, &envp, &file_actions, &current)?;
+
+    unsafe { *child_pid_ptr = child_pid };
+    Ok(0)
+}
+
+#[repr(C)]
+pub struct SpawnFileActions {
+    allocated: u32,
+    used: u32,
+    actions: *const SpawnAction,
+    pad: [u32; 16],
+}
+
+#[repr(C)]
+struct SpawnAction {
+    tag: u32,
+    action: Action,
+}
+
+impl SpawnAction {
+    // Wrappers to access the union field safely
```

Comment:
I admire the good intention of defining these wrapper functions. But I am afraid that they are not as useful as they are intended to be.

First, they make the code unnecessarily complex. And defining each of the wrapper functions takes four lines of code, but they are all essentially one-liner and they are only used once. So defining the wrapper functions does not deliver any benefit  of code reuse.

Second, the `assert` statements are pointless. Because each wrapper is only used once and used in the correct branches of `match sa.tag`. So they cannot deliver any real benefit of safety.

I recommend to rewrite as followinng:

```rust
#[repr(C)]
struct SpawnAction {
    tag: u32,
    action: Action,
}

impl SpawnAction {
    pub fn to_file_action(&self) -> Result<FileAction> {
        match self.tag {
            SPAWN_DO_CLOSE => FileAction::Close(unsafe { self.action.close_action.fd }),
            SPAWN_DO_DUP2 => FileAction::Dup2(unsafe { ... }),
            SPAWN_DO_OPEN => FileAction::Open {
                path: clone_cstring_safely(unsafe { ... })?
                    .to_string_lossy()
                    .into_owned(),
                mode: unsafe { ... },
                oflag: unsafe { ... },
                fd: unsafe { ... },
            },
            _ => return_errno!(EINVAL, "Unknown file action tag"),
        };
    }
}
```

The above version removes all (useless) wrappers without sacrificing memory safety. And `clone_file_actions_from_fa_safely` looks more neat.

Reply:
The reason why I choose to add wrapper functions is as follows:
The wrapper functions can ensure the unsafe value can be accessed safely, if users access the `SpawnAction` by mistake, it can panic.

The `to_file_action` in your suggestion looks simpler, good idea